### PR TITLE
feat!: throw exception if webpack config file exist and running with Vite

### LINKED ISSUE #14417: Prevent application startup if webpack config file is present while running with Vite
### Describe your motivation
As of V23.2, the default frontend build tool is changed to Vite (and webpack is only available behind a feature flag). 
Applications migrating from earlier versions normally have the legacy `webpack.config.js`,  should be notified about the fact that any webpack configuration will not be picked by Vite. 

### Describe the solution you'd like

To prevent confusion and misconfiguration, a runtime exception should prevent application startup if `webpack.config.js` exists in the project and it contents are different from the default that is created by Vaadin. The exception message should clearly ask the user to either:
- Migrate any custom webpack configs to Vite alternatives and remove the `webpack.config.js` to be able to run the application with Vite.
- Reactivate the webpack with its associated feature flag to continue running the application with webpack.

If the `webpack.config.js` exists but the contents were the autogenerated content by Vaadin, only WARN the user about its existence and don't prevent the application startup.

### PR DESCRIPTION
**Pull request by taefi**
## Description

Throw exception if webpack.config.js file 
with custom configurations exists in the 
project and running with Vite.
If the default webpack.config.js exists, 
just warn the user about its existence but 
do not prevent the application startup.

Fixes #14417

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed a self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [x] Enhancement / new feature was discussed in internal slack.

### CODE REVIEW

**Comment by knoobie** [comment]
I would argue that this is a huge DX-downer, resulting in every project updating from < 23.2 to fail with an exception even tho only the default (generated webpack file by Vaadin!) which is probably the default for 70-90% of all flow developer.

Ist there a way to reliable detect e.g. with a simple checksum check that's the Default content in the file and just "migrate" to vite? Instead of forcing the developer to delete the file by hand?

**Comment by mcollovati** [comment]
I think we should throw only if a custom web pack.config.js is present. If it is the default one provided by Vaadin we can just ignore it or add a log message stating that webpack files can be safely deleted

**Comment by mcollovati** [comment]
Perhaps we should also skip the check if webpack feature flag is set to true

**Comment by mshabarov** [comment]
I agree with the comments above.
We supposed to have a __custom__ webpack config detection, not just a webpack config file detection.
Since it is being generated by Vaadin for all the projects, it presents always, so this detection would trigger always.
The detection should check that the config file contains non-empty object to merge, or it has an extra codes for handling `flowDefaults`.
Here is an example of webpack config in starter project:
https://github.com/vaadin/skeleton-starter-flow-spring/blob/v23/webpack.config.js
(however it is slightly different from the template in the flow-server).

Detection should somehow extract the meaningful part of the config: either second parameter object in `merge` or an extra codes apart from `module.exports = merge(flowDefaults, {})`

**Comment by taefi** [comment]
> Perhaps we should also skip the check if webpack feature flag is set to true

I think this one has already been taken care of, as the validation task is only added if Vite is in use.

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskNotifyWebpackConfExistenceWhileUsingViteTest.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            TaskNotifyWebpackConfExistenceWhileUsingVite task = new TaskNotifyWebpackConfExistenceWhileUsingVite(
+                    new File(userDir));
+            task.execute();
+
+            Mockito.verify(logger, times(1)).warn(anyString());
+        }
+    }
+
+    @Test
+    public void should_executeWithoutWarning_when_webpackConfigFileDoesNotExist() {
```
Comment (mshabarov) [inline flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskNotifyWebpackConfExistenceWhileUsingViteTest.java:87]: There should be also a test checking that no warning message is present if webpack is used.

**Review by mshabarov — CHANGES_REQUESTED** [review]

File: flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskNotifyWebpackConfExistenceWhileUsingViteTest.java (hunk 1 above)
Comment (taefi) [reply flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskNotifyWebpackConfExistenceWhileUsingViteTest.java]: Verifying the log message for this specific case needs a very complicated mocking for a lot of static methods, as in this case, the NodeTask execution will not fail and the next steps would result in errors that are not in the scope of this test. Instead, another test created in `NodeTasksWebpackTest` to ensure that the `TaskNotifyWebpackConfExistenceWhileUsingVite` is not even added to the list of fallible commands when `webpack` is activated via its feature flag, so no warn or exception would happen in this case.

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskNotifyWebpackConfExistenceWhileUsingVite.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                "merge(flowDefaults,);");
+    }
+
+    private String removeWhiteSpaces(String content) {
+        return content.replaceAll("\\s", "");
+    }
+
+    private String removeComments(String content) {
+        return content.replaceAll(
+                "(?:/\\*(?:[^*]|(?:\\*+[^*/]))*\\*+/)|(?://.*)", "");
```
[the remark this answers is not in the document]
Comment (taefi) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskNotifyWebpackConfExistenceWhileUsingVite.java:168]: The existing test was modified to have different types of possible comments in the mock `webpack.config.js` content and it is verified to be working correctly as it still detects the config file as the default generated one.

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskNotifyWebpackConfExistenceWhileUsingVite.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            return;
+        }
+
+        long projectWebpackConfigContentHash = calculateProjectWebpackConfigContentHashCode(
+                webpackConfigFilePath);
+        long templateWebpackConfigContentHash = calculateTemplateWebpackConfigContentHashCode();
+
+        if (projectWebpackConfigContentHash != templateWebpackConfigContentHash) {
+            // custom content exists in the project's webpack.config.js file:
+            throw new IllegalStateException(String.format(ERROR_MESSAGE));
```
[the remark this answers is not in the document]
Comment (taefi) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskNotifyWebpackConfExistenceWhileUsingVite.java]: Done.

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskNotifyWebpackConfExistenceWhileUsingVite.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.vaadin.flow.server.frontend.FrontendUtils.WEBPACK_CONFIG;
+
+/**
+ * Notifies the user about the existence of webpack.config.js while the project
+ * is running Vite as the frontend build tool. This can be helpful especially
+ * when migrating to 23.2 and later since it prevent any confusion or any
+ * accidental misconfiguration in webpack related config files while using vite.
```
[the remark this answers is not in the document]
Comment (taefi) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskNotifyWebpackConfExistenceWhileUsingVite.java]: Done.

**Comment by mshabarov** [comment]
I've tested with starter project with webpack configuration:
- default webpack config - > warning shown
- custom webpack config -> app stopped with message
- renamed config -> starts with no warnings
- custom config + reactivated webpack -> starts with no warnings.

Works as expected.

**Pull request merged**